### PR TITLE
Added initial version of account-level installer

### LINKED ISSUE #920: [FEATURE]: Account level installer
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

Currently we install UCX to a single workspace at a time. This is unsustainable for accounts with 10s and 100s of workspaces.

### Proposed Solution

Allow admins to run the installer as account admins and install it to all the workspaces in a given account at once.

- Perform CLI auth to login to the account.
- Run the installer.
- Select the option to install to all workspaces
- UCX is installed to all workspaces.
- We place a `config.yml` with a list of all workspace ids
- we create workspaces.json with workspace id to workspace name mapping 

### Additional Context
Related to:
- #918
- https://github.com/databrickslabs/ucx/issues/919
- https://github.com/databrickslabs/ucx/issues/891
- https://github.com/databrickslabs/ucx/issues/673
- https://github.com/databrickslabs/ucx/issues/342
- https://github.com/databrickslabs/ucx/issues/335

### PR DESCRIPTION
**Pull request by nkvuong**
## Changes

Allow admins to run the installer as account admins and install it to all the workspaces in a given account at once.

- Perform necessary auth to login to the account.
- Prompt for configuration for the first workspace
- Run the installer.
- Confirm that user wants to repeat that for the remaining workspaces
- Repeat
- UCX is installed to all workspaces.

This is done by splitting the prompts for `_configure_new_installation` into a new method `prompt_for_new_installation`, which saves the answers to a new `InstallationConfig` data class. This allows the answers to be reused for other workspaces

### Linked issues

Resolves #920

### Functionality 

- [x] added relevant user documentation
- [x] modified existing command: `databricks labs install ucx ...` will switch to account-level installation when the environment variable `UCX_FORCE_INSTALL` is set to `account`

### Tests

- [x] manually tested

### CODE REVIEW

File: src/databricks/labs/ucx/config.py [hunk 1]
```diff
@@ -66,3 +68,22 @@ def v1_migrate(cls, raw: dict) -> dict:
...
         raw["version"] = 2
         return raw
+
+
+@dataclass
+class InstallationConfig:
+    # Installation parameters
+    silent: bool
+    inventory_database: str
+    configure_groups: ConfigureGroups
```
[the remark this answers is not in the document]
Comment (nkvuong) [reply src/databricks/labs/ucx/config.py]: fixed

File: src/databricks/labs/ucx/config.py [hunk 2]
```diff
@@ -66,3 +68,22 @@ def v1_migrate(cls, raw: dict) -> dict:
         raw["renamed_group_prefix"] = groups.get("backup_group_prefix", "db-temp-")
         raw["version"] = 2
         return raw
+
+
+@dataclass
+class InstallationConfig:
```
[the remark this answers is not in the document]
Comment (nkvuong) [reply src/databricks/labs/ucx/config.py]: fixed

File: src/databricks/labs/ucx/install.py [hunk 3]
```diff
@@ -538,15 +556,93 @@ def _trigger_workflow(self, step: str):
...
     installer = WorkspaceInstaller(prmpts, current, workspace_client, app)
-    installer.run()
+    installer.run(installer.prompt_for_new_installation())
+
+
+if __name__ == "__main__":
+    logger = get_logger(__file__)
+    logger.setLevel("INFO")
+    app = ProductInfo.from_class(WorkspaceConfig)
+    prmpts = Prompts()
```
[the remark this answers is not in the document]
Comment (nkvuong) [reply src/databricks/labs/ucx/install.py]: fixed

File: src/databricks/labs/ucx/install.py [hunk 4]
```diff
@@ -538,15 +556,93 @@ def _trigger_workflow(self, step: str):
...
+
+        try:
+            current = app.current_installation(workspace_client)
+        except NotFound:
+            current = Installation.assume_global(workspace_client, app.product_name())
+        installer = WorkspaceInstaller(prmpts, current, workspace_client, app)
+        if not confirmed or not installation_config:
+            installation_config = installer.prompt_for_new_installation(True)
+        try:
+            installer.run(installation_config)
```
[the remark this answers is not in the document]
Comment (nkvuong) [reply src/databricks/labs/ucx/install.py]: WorkspaceInstallation depends on the current installation, which is workspace-specific so we cannot reuse them without some significant refactoring. It is the same with WheelsV2 as it requires a workspace client

File: src/databricks/labs/ucx/install.py [hunk 5]
```diff
@@ -538,15 +556,93 @@ def _trigger_workflow(self, step: str):
...
         current = Installation.assume_global(workspace_client, app.product_name())
-    prmpts = Prompts()
     installer = WorkspaceInstaller(prmpts, current, workspace_client, app)
-    installer.run()
+    installer.run(installer.prompt_for_new_installation())
+
+
+if __name__ == "__main__":
+    logger = get_logger(__file__)
+    logger.setLevel("INFO")
```
[the remark this answers is not in the document]
Comment (nkvuong) [reply src/databricks/labs/ucx/install.py]: this was in the existing codebase, removed

File: src/databricks/labs/ucx/install.py [hunk 6]
```diff
@@ -538,15 +556,93 @@ def _trigger_workflow(self, step: str):
...
+        w = WorkspaceClient(product="ucx", product_version=__version__)
+        if w.config.is_aws:
+            host = "https://accounts.cloud.databricks.com"
+        elif w.config.is_azure:
+            host = "https://accounts.azuredatabricks.net"
+        elif w.config.is_gcp:
+            host = "https://accounts.gcp.databricks.com"
+        else:
+            raise ValueError("Unknown cloud provider")
+        a = AccountClient(host=host, product="ucx", product_version=__version__)
```
[the remark this answers is not in the document]
Comment (nkvuong) [reply src/databricks/labs/ucx/install.py]: done

File: src/databricks/labs/ucx/install.py [hunk 7]
```diff
@@ -538,15 +556,93 @@ def _trigger_workflow(self, step: str):
...
+    return a
+
+
+def _is_valid_workspace(workspace_client: WorkspaceClient):
+    try:
+        # check if user is a workspace admin
+        current_user = workspace_client.current_user.me()
```
[the remark this answers is not in the document]
Comment (nkvuong) [reply src/databricks/labs/ucx/install.py]: pylint does not like short variable name

File: src/databricks/labs/ucx/install.py [hunk 8]
```diff
@@ -538,15 +556,93 @@ def _trigger_workflow(self, step: str):
...
+            installer.run(installation_config)
+        except RuntimeWarning:
+            logger.info("Skipping workspace...")
+        # if user confirms to install on remaining workspaces with the same config, reuse the object
+        if confirmed:
+            continue
+        confirmed = prmpts.confirm("Do you want to install UCX on the remaining workspaces with the same config?")
+
+    # upload the json dump of workspace info in the .ucx folder
+    ctx.account_workspaces.sync_workspace_info(installed_workspaces)
```
[the remark this answers is not in the document]
Comment (nkvuong) [reply src/databricks/labs/ucx/install.py]: done

File: src/databricks/labs/ucx/install.py [hunk 9]
```diff
@@ -169,11 +176,43 @@ def run(
             self._product_info,
         )
         try:
-            workspace_installation.run()
+            workspace_installation.run(self._is_account_install, config.trigger_job)
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:179]:
```suggestion
            workspace_installation.run()
```
Config is already injected via constructor, inject the property about account-level install

File: src/databricks/labs/ucx/install.py [hunk 10]
```diff
@@ -233,58 +272,43 @@ def _apply_upgrades(self):
...
-            num_threads=num_threads,
             instance_profile=instance_profile,
             spark_conf=spark_conf_dict,
             min_workers=min_workers,
             max_workers=max_workers,
             policy_id=policy_id,
             instance_pool_id=instance_pool_id,
-            is_terraform_used=is_terraform_used,
-            include_databases=self._select_databases(),
+            installed_workspace_ids=self._installed_workspace_ids,
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:298]: Shouldn't we add a list of workspaces only after we successfully installed on a particular workspace? That way it's clear that operation succeeded or not, preventing downstream bugs

File: src/databricks/labs/ucx/install.py [hunk 11]
```diff
@@ -415,8 +440,7 @@ def config(self):
     def folder(self):
         return self._installation.install_folder()
 
-    def run(self):
-        logger.info(f"Installing UCX v{self._product_info.version()}")
+    def run(self, silent: bool = False, trigger_job: bool = False):
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:443]:
```suggestion
    def run(self, trigger_job: bool = False):
```
If we have accounts install, why do we need silent?

File: src/databricks/labs/ucx/install.py [hunk 12]
```diff
@@ -538,15 +562,107 @@ def _trigger_workflow(self, step: str):
...
+        return accessible_workspaces
+
+    def install_on_account(self, app: ProductInfo | None = None):
+        ctx = AccountContext(self._get_safe_account_client())
+        if app is None:
+            app = ProductInfo.from_class(WorkspaceConfig)
+        default_config = None
+        confirmed = False
+        accessible_workspaces = self._get_accessible_workspaces()
+        msg = "\n".join([f"{w.deployment_name}" for w in accessible_workspaces])
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:612]: Why do we need f-string?

File: src/databricks/labs/ucx/install.py [hunk 13]
```diff
@@ -538,15 +562,107 @@ def _trigger_workflow(self, step: str):
...
+        installed_workspace_ids = [w.workspace_id for w in accessible_workspaces if w.workspace_id is not None]
+        if not self.prompts.confirm(
+            f"UCX has detected the following workspaces available to install. \n{msg}\n" f"Do you want to continue?"
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:615]:
```suggestion
            f"UCX has detected the following workspaces available to install. \n{msg}\nDo you want to continue?"
```

File: tests/unit/test_install.py [hunk 14]
```diff
@@ -1262,6 +1264,7 @@ def test_runs_upgrades_on_too_old_version(ws, any_prompt):
     sql_backend = MockBackend()
     wheels = create_autospec(WheelsV2)
     install.run(
+        default_config=WorkspaceConfig(inventory_database='ucx'),
```
Comment (nfx) [inline tests/unit/test_install.py:1267]: Can you rewrite the logic the way that existing tests won't have to be modified to keep working with the new functionality? 

There's a high probability of hidden bugs if changes here are required.

File: tests/unit/test_install.py [hunk 15]
```diff
@@ -1638,3 +1650,23 @@ def test_extract_major_minor_versions():
 
     version_string3 = "should not match"
     assert extract_major_minor(version_string3) is None
+
+
+def test_account_installer(ws):
+    acc = create_autospec(AccountClient)
+    acc.workspaces.list.return_value = [Workspace(workspace_id=123), Workspace(workspace_id=456)]
```
Comment (nfx) [inline tests/unit/test_install.py:1657]: Aren't missing workspace names causing problems?

File: src/databricks/labs/ucx/account.py [hunk 16]
```diff
@@ -38,29 +38,33 @@ def _get_cloud(self) -> str:
...
         """
         clients = []
-        for workspace in self._workspaces():
+        for workspace in workspaces:
             ws = self.client_for(workspace)
             clients.append(ws)
         return clients
 
-    def sync_workspace_info(self):
+    def sync_workspace_info(self, workspaces: list[Workspace] | None = None):
```
Comment (nfx) [inline src/databricks/labs/ucx/account.py:52]: Does the cli command still work? I'm surprised that no test failed on this change to public api

**Review by nfx — CHANGES_REQUESTED** [review]
Can you rewrite the logic the way that existing tests won't have to be modified to keep working with the new functionality? There's a high probability of hidden bugs if changes here are required.

File: src/databricks/labs/ucx/account.py (hunk 16 above)
Comment (nkvuong) [reply src/databricks/labs/ucx/account.py:52]: the new param is optional, so does not affect existing behaviour

File: tests/unit/test_install.py (hunk 15 above)
Comment (nkvuong) [reply tests/unit/test_install.py:1650]: the workspace names are only used for the prompt message, so the lack of it does not cause issue

File: src/databricks/labs/ucx/account.py [hunk 17]
```diff
@@ -38,29 +38,33 @@ def _get_cloud(self) -> str:
     def client_for(self, workspace: Workspace) -> WorkspaceClient:
         return self._ac.get_workspace_client(workspace)
 
-    def workspace_clients(self) -> list[WorkspaceClient]:
+    def workspace_clients(self, workspaces: list[Workspace]) -> list[WorkspaceClient]:
```
[the remark this answers is not in the document]
Comment (nkvuong) [reply src/databricks/labs/ucx/account.py]: make new param optional

**Review by nfx — APPROVED** [review]
lgtm

File: src/databricks/labs/ucx/install.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nfx) [reply src/databricks/labs/ucx/install.py]: oh, need to add a whitelist

**Pull request merged**